crates/air-lib/interpreter-data/src/interpreter_data/verification.rs

```diff
@@ -0,0 +1,284 @@
+        let mut grouped_cids: HashMap<Box<str>, PeerInfo<'data>> = data
+            .signatures
+            .iter()
+            .map(|(public_key, signature)| {
+                (
+                    public_key.to_peer_id().to_string().into(),
+                    PeerInfo::new(public_key, signature),
+                )
+            })
+            .collect();
+
+        // fill PeerInfo's `cids` field, checking for peer IDs without a key
+        collect_peers_cids_from_trace(&data.trace, &data.cid_info, &mut grouped_cids)?;
+
+        // sort cids for canonicalization
+        for peer_info in grouped_cids.values_mut() {
+            peer_info.cids.sort_unstable();
+        }
+
+        Ok(Self {
+            grouped_cids,
+            particle_id,
+        })
+    }
+
+    /// Verify each peers' signatures.
+    pub fn verify(&self) -> Result<(), DataVerifierError> {
+        for peer_info in self.grouped_cids.values() {
+            peer_info
+                .public_key
+                .verify(&peer_info.cids, self.particle_id, peer_info.signature)
+                .map_err(|error| DataVerifierError::SignatureMismatch {
+                    error: error.into(),
+                    cids: peer_info.cids.clone(),
+                    peer_id: peer_info.public_key.to_peer_id().to_string(),
+                })?;
+        }
+        Ok(())
+    }
+
+    /// For each peer, merge previous and current CID multisets by determining the largest set.
+    ///
+    /// This code uses an invariant: peer's multiset of produced CIDs is always a superset of
+    /// previous invocation's multiset:
+    ///
+    /// A_0 ⊆ A_1 ⊆ ... ⊆ A_n.
+    ///
+    /// So, the largest multiset is selected as the result of merging, the invariant is checked,
+    /// and a error is returned if it is violated.
+    ///
+    /// If the multisets are of same size, they have to be equal.
+    // TODO enforce merging only verified sets
+    pub fn merge(mut self, other: Self) -> Result<SignatureStore, DataVerifierError> {
+        use std::collections::hash_map::Entry::*;
+
+        for (other_peer_pk, mut other_info) in other.grouped_cids {
+            let our_info = self.grouped_cids.entry(other_peer_pk);
+            match our_info {
+                Occupied(mut our_info_ent) => {
+                    debug_assert_eq!(other_info.public_key, our_info_ent.get().public_key);
+
+                    if our_info_ent.get().cids.len() < other_info.cids.len() {
+                        // the merged map countains largest set for each peer_id
+                        //
+                        // this code assumes that a peer only adds CIDs to its set, so CID multisets
+                        //   are growing-only; but it is additionally checked below
+                        // so, we get a largest set as merged one
+                        std::mem::swap(our_info_ent.get_mut(), &mut other_info);
+                    }
```

Comment:
I think that this code is so important that it deserve to be placed in a separate function

Reply:
That function mutates its arguments; it looks ugly/unclear no matter how I refactor.